### PR TITLE
Use enum class to define protocols message type for strong typing

### LINKED ISSUE #4592: Use enum class to define protocols message type for type safty
#### Problem
This is follow-up of https://github.com/project-chip/connectedhomeip/pull/4429, we already have a set of APIs that allow us to just pass in a strongly typed (enum class) message type and derive the protocol from the type of the message type.

 #### Proposed Solution
update all protocol message type to enum class

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
This is follow-up of #4429 which defined a set of APIs that allow us to just pass in a strongly typed (enum class) message type and derive the protocol from the type of the message type. Currently, only SecureChannel protocol has been updated to use strongly typed (enum class) message type.

 #### Summary of Changes
update all protocol message type to use strongly typed (enum class) message type.

 Fixes #4592

### CODE REVIEW

**Review by saurabhst — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]
Thank you!

**Review by bzbarsky-apple — APPROVED** [review]
Thank you!

File: src/messaging/ExchangeMgr.h [hunk 1]
```diff
@@ -107,7 +107,7 @@ class DLL_EXPORT ExchangeManager : public SecureSessionMgrDelegate
      *                                                             is full and a new one cannot be allocated.
      *  @retval #CHIP_NO_ERROR On success.
      */
-    CHIP_ERROR RegisterUnsolicitedMessageHandler(uint32_t protocolId, ExchangeDelegate * delegate);
+    CHIP_ERROR RegisterUnsolicitedMessagesHandler(uint32_t protocolId, ExchangeDelegate * delegate);
```
Comment (pan-apple) [inline src/messaging/ExchangeMgr.h:110]: Is the plural `Messages` intentional change?

**Review by pan-apple — APPROVED** [review]

File: src/messaging/ExchangeMgr.h (hunk 1 above)
Comment (yufengwangca) [reply src/messaging/ExchangeMgr.h:110]: Yes, there are two reasons:

1. If we overload the following two functions:
    CHIP_ERROR RegisterUnsolicitedMessageHandler(uint32_t protocolId, ExchangeDelegate * delegate);
    CHIP_ERROR RegisterUnsolicitedMessageHandler(MessageType msgType, ExchangeDelegate * delegate);

err = mpExchangeMgr->RegisterUnsolicitedMessagesHandler(Protocols::kProtocol_InteractionModel, this); will confuse, and call into second one from my test, which is not what we want.

2. CHIP_ERROR RegisterUnsolicitedMessagesHandler(uint32_t protocolId, ExchangeDelegate * delegate); is to register a handler for a group of messages, this should be reflected in the function name.

File: src/messaging/ExchangeMgr.h [hunk 2]
```diff
@@ -148,6 +159,17 @@ class DLL_EXPORT ExchangeManager : public SecureSessionMgrDelegate
      */
     CHIP_ERROR UnregisterUnsolicitedMessageHandler(uint32_t protocolId, uint8_t msgType);
 
+    /**
+     * A strongly-message-typed version of UnregisterUnsolicitedMessageHandler.
+     */
+    template <typename MessageType, typename = std::enable_if_t<std::is_enum<MessageType>::value>>
+    CHIP_ERROR UnregisterUnsolicitedMessageHandler1(MessageType msgType)
```
Comment (andy31415) [inline src/messaging/ExchangeMgr.h:166]: The 1 suffix is a bit ugly.

How about something like:

`UnregisteerUnsolicitedMessagesHandlerForProtodcol` and `UnregisterUnsolicitedMessageHandlerForType`

Overloads for non-equivalent types are very hard to follow.

**Review by andy31415 — APPROVED** [review]

File: src/messaging/ExchangeMgr.cpp [hunk 3]
```diff
@@ -107,22 +107,23 @@ ExchangeContext * ExchangeManager::NewContext(SecureSessionHandle session, Excha
...
 
-CHIP_ERROR ExchangeManager::RegisterUnsolicitedMessageHandler(uint32_t protocolId, uint8_t msgType, ExchangeDelegate * delegate)
+CHIP_ERROR ExchangeManager::RegisterUnsolicitedMessageHandlerForType(uint32_t protocolId, uint8_t msgType,
+                                                                     ExchangeDelegate * delegate)
 {
     return RegisterUMH(protocolId, static_cast<int16_t>(msgType), delegate);
 }
 
-CHIP_ERROR ExchangeManager::UnregisterUnsolicitedMessageHandler(uint32_t protocolId)
+CHIP_ERROR ExchangeManager::UnregisteerUnsolicitedMessageHandlerForProtodcol(uint32_t protocolId)
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeMgr.cpp:121]:
```suggestion
CHIP_ERROR ExchangeManager::UnregisterUnsolicitedMessageHandlerForProtocol(uint32_t protocolId)
```

File: src/messaging/ExchangeMgr.h [hunk 4]
```diff
@@ -133,7 +144,7 @@ class DLL_EXPORT ExchangeManager : public SecureSessionMgrDelegate
      *                                                       is not found.
      *  @retval #CHIP_NO_ERROR On success.
      */
-    CHIP_ERROR UnregisterUnsolicitedMessageHandler(uint32_t protocolId);
+    CHIP_ERROR UnregisteerUnsolicitedMessageHandlerForProtodcol(uint32_t protocolId);
```
Comment (bzbarsky-apple) [inline src/messaging/ExchangeMgr.h:147]:
```suggestion
    CHIP_ERROR UnregisterUnsolicitedMessageHandlerForProtocol(uint32_t protocolId);
```

File: src/messaging/tests/TestExchangeMgr.cpp [hunk 5]
```diff
@@ -110,22 +110,22 @@ void CheckUmhRegistrationTest(nlTestSuite * inSuite, void * inContext)
...
-    err = ctx.GetExchangeManager().RegisterUnsolicitedMessageHandler(0x0002, 0x0001, &mockAppDelegate);
+    err = ctx.GetExchangeManager().RegisterUnsolicitedMessageHandlerForType(0x0002, 0x0001, &mockAppDelegate);
     NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
 
-    err = ctx.GetExchangeManager().UnregisterUnsolicitedMessageHandler(0x0001);
+    err = ctx.GetExchangeManager().UnregisteerUnsolicitedMessageHandlerForProtodcol(0x0001);
     NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
 
-    err = ctx.GetExchangeManager().UnregisterUnsolicitedMessageHandler(0x0002);
+    err = ctx.GetExchangeManager().UnregisteerUnsolicitedMessageHandlerForProtodcol(0x0002);
```
Comment (bzbarsky-apple) [inline src/messaging/tests/TestExchangeMgr.cpp:122]:
```suggestion
    err = ctx.GetExchangeManager().UnregisterUnsolicitedMessageHandlerForProtocol(0x0002);
```

File: src/messaging/tests/TestExchangeMgr.cpp [hunk 6]
```diff
@@ -110,22 +110,22 @@ void CheckUmhRegistrationTest(nlTestSuite * inSuite, void * inContext)
...
-    err = ctx.GetExchangeManager().RegisterUnsolicitedMessageHandler(0x0001, &mockAppDelegate);
+    err = ctx.GetExchangeManager().RegisterUnsolicitedMessageHandlerForProtocol(0x0001, &mockAppDelegate);
     NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
 
-    err = ctx.GetExchangeManager().RegisterUnsolicitedMessageHandler(0x0002, 0x0001, &mockAppDelegate);
+    err = ctx.GetExchangeManager().RegisterUnsolicitedMessageHandlerForType(0x0002, 0x0001, &mockAppDelegate);
     NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
 
-    err = ctx.GetExchangeManager().UnregisterUnsolicitedMessageHandler(0x0001);
+    err = ctx.GetExchangeManager().UnregisteerUnsolicitedMessageHandlerForProtodcol(0x0001);
```
Comment (bzbarsky-apple) [inline src/messaging/tests/TestExchangeMgr.cpp:119]:
```suggestion
    err = ctx.GetExchangeManager().UnregisterUnsolicitedMessageHandlerForProtocol(0x0001);
```

**Review by bzbarsky-apple — DISMISSED** [review]

File: src/messaging/tests/TestExchangeMgr.cpp (hunk 6 above)
Comment (yufengwangca) [reply src/messaging/tests/TestExchangeMgr.cpp]: Thank you!

**Review by woody-apple — APPROVED** [review]
Approving given @pan-apple 's approval
